implementations/rust/ockam/ockam_macro/src/lib.rs

```diff
@@ -0,0 +1,179 @@
+        }
+        _ => None,
+    }
+}
+// Gets the inner of a type SomeType<Inner> or none if it doesn't exist
+fn get_inner(ty: &Type) -> Option<&Type> {
+    match ty {
+        Type::Path(tp) if tp.qself.is_none() => {
+            let mut tp = tp
+                .path
+                .segments
+                .iter()
+                .skip_while(|s| s.arguments.is_empty());
+            if let Some(segment) = tp.next() {
+                match &segment.arguments {
+                    syn::PathArguments::AngleBracketed(ab) if ab.args.len() == 1 => {
+                        if let Some(syn::GenericArgument::Type(t)) = ab.args.first() {
+                            return Some(t);
+                        }
+                        return None;
+                    }
+                    _ => return None,
+                }
+            }
+            None
+        }
+        _ => None,
+    }
+}
+
+fn has_generic(ty: &Type, generics_list: &[&Ident]) -> bool {
+    if let Some(inner) = get_inner(ty) {
+        return has_generic(inner, generics_list);
+    }
+    if let Type::Path(tp) = ty {
+        if generics_list.contains(&&tp.path.segments[0].ident) {
+            return true;
+        }
+    }
+    false
+}
+
+#[proc_macro_derive(AsyncTryClone)]
+pub fn derive(input: TokenStream) -> TokenStream {
+    let ast = parse_macro_input!(input as DeriveInput);
+    let struct_data = match ast.data {
+        Struct(s) => s,
+        _ => {
+            panic!("currently only works for structs");
+        }
+    };
+    let struct_ident = ast.ident;
+    let named_fields = match struct_data.fields {
+        syn::Fields::Named(f) => f.named,
+        _ => {
+            panic!("currently only works for named fields");
+        }
+    };
+
+    // Get generic type params from struct definition
+    let generic_tys = ast
+        .generics
+        .type_params()
+        .map(|t| &t.ident)
+        .collect::<Vec<_>>();
+
+    // Types for form name: T where T is a generic type
+    let simple_generic_fields = named_fields
+        .iter()
+        .filter_map(|f| {
+            let outer = get_outer(&f.ty).unwrap();
+            if generic_tys.iter().any(|id| id.to_string() == outer) {
+                return Some(outer);
+            }
+            None
+        })
+        .collect::<Vec<_>>();
+
+    // Types which have a generic and are not simple
+    let complex_generic_fields = named_fields
+        .iter()
+        .filter_map(|f| {
+            if has_generic(&f.ty, &generic_tys) && get_inner(&f.ty).is_some() {
+                return Some(&f.ty);
+            }
+            None
+        })
+        .collect::<Vec<_>>();
+
+    let mut generics = ast.generics;
+
+    // Add trait bounds on generic type params
+    for p in &mut generics.params {
+        if let GenericParam::Type(ref mut t) = *p {
+            // Every generic type must be Send + Sync
+            t.bounds.push(parse_quote!(::std::marker::Send));
```

Comment:
Let's import it from core::marker for no_std builds

Reply:
Done